### PR TITLE
[EOS-19647] : BG Services Integration with IEM framework infrastructure

### PR DESCRIPTION
**Pull request by pranavPawar07**
Signed-off-by: Pranav Diliprao Pawar <pranav.pawar@seagate.com>

### CODE REVIEW

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 1]
```diff
@@ -21,24 +21,27 @@
...
+    module = ""
 
     # List of eventcodes
     S3_CONN_FAILURE = "0050010001"
 
     # List of eventstrings
     S3_CONN_FAILURE_STR = "Failed to connect to S3 server. For more information refer the Troubleshooting Guide."
 
-    def __init__(self, loglevel, eventcode, eventstring):
+    def __init__(self, module, loglevel, eventcode, eventstring):
```
Comment (vimalraghwani20) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:40]: @nileshgovande 
can we hard code module as S3 BG delete?

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 2]
```diff
@@ -49,5 +52,32 @@ def __init__(self, loglevel, eventcode, eventstring):
...
+
+    @classmethod
+    def receive(self, component):
+        """Receive the incoming message."""
+        EventMessage.subscribe(component)
+
+        try:
+            alert = EventMessage.receive()
+        except Exception as exception:
+            msg = ("iem except:%s %s") % (
```
Comment (vimalraghwani20) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:80]: No need to return msg. 
see of we can log exception and continue..

File: s3backgrounddelete/ut/cortx_s3_iem_tests.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+from s3backgrounddelete.IEMutil import IEMutil
+
+def test_send():
```
Comment (vimalraghwani20) [inline s3backgrounddelete/ut/cortx_s3_iem_tests.py:22]: @nileshgovande  to run this test code, we need configure-iem.sh script which i have created as part of init.sh 

If we dont want to integrate this script then we should remove the UTs

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 4]
```diff
@@ -49,5 +52,32 @@ def __init__(self, loglevel, eventcode, eventstring):
...
+    def send(self, module, event_id, severity, message_blob):
+        """Send the message."""
+        source = 'S'
+        component= 'S3'
+        EventMessage.init(component, source)
+
+        try:
+            EventMessage.send(module=module, event_id=event_id, severity=severity, message_blob=message_blob)
+        except Exception as exception:
+            msg = ("iem except:%s %s") % (
```
Comment (vimalraghwani20) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:67]: ignore exception and continue

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 1 above)
Comment (nileshgovande) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py:40]: have default value - OK to pass

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 5]
```diff
@@ -49,5 +52,32 @@ def __init__(self, loglevel, eventcode, eventstring):
...
-        IEM = "IEC:%sS%s:%s" % (self.severity, self.eventCode, self.eventString)
-        syslog.syslog(IEM)
+        self.send(self.module, self.eventCode, self.severity, self.eventString)
+
+    @classmethod
+    def send(self, module, event_id, severity, message_blob):
+        """Send the message."""
+        source = 'S'
+        component= 'S3'
+        EventMessage.init(component, source)
```
Comment (nileshgovande) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:62]: Is it required for every time ? - if no move to __init()__

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 4 above)
Comment (nileshgovande) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py:67]: Ignore exception but add a log message
Take logger as input to constructor.

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 2 above)
Comment (nileshgovande) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py:80]: Ignore exception but add a log message
Take logger as input to constructor.

File: s3backgrounddelete/ut/cortx_s3_iem_tests.py (hunk 3 above)
Comment (nileshgovande) [reply s3backgrounddelete/ut/cortx_s3_iem_tests.py:22]: Check for the existance of config file and if present, continue

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 1 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 2 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done

File: s3backgrounddelete/ut/cortx_s3_iem_tests.py (hunk 3 above)
Comment (pranavPawar07) [reply s3backgrounddelete/ut/cortx_s3_iem_tests.py:25]: done

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 4 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 5 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 6]
```diff
@@ -49,5 +53,27 @@ def __init__(self, loglevel, eventcode, eventstring):
...
-        syslog.syslog(IEM)
+        self.send('S3 BG delete', self.eventCode, self.severity, self.eventString)
+
+    @classmethod
+    def send(self, module, event_id, severity, message_blob):
+        """Send the message."""
+        try:
+            EventMessage.send(module='S3 BG delete', event_id=event_id, severity=severity, message_blob=message_blob)
+        except:
+            self._logger.ERROR("Error in sending the IEM message.")
```
Comment (vimalraghwani20) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:64]: log the exception.

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py [hunk 7]
```diff
@@ -49,5 +53,27 @@ def __init__(self, loglevel, eventcode, eventstring):
...
+    @classmethod
+    def receive(self, component):
+        """Receive the incoming message."""
+        EventMessage.subscribe(component)
+
+        try:
+            alert = EventMessage.receive()
+            self._logger.DEBUG("Received message --> ", alert)
+        except:
+            self._logger.ERROR("Error in receiving the IEM message.")
```
Comment (vimalraghwani20) [inline s3backgrounddelete/s3backgrounddelete/IEMutil.py:77]: log the exception

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 6 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done. please check

File: s3backgrounddelete/s3backgrounddelete/IEMutil.py (hunk 7 above)
Comment (pranavPawar07) [reply s3backgrounddelete/s3backgrounddelete/IEMutil.py]: done please check

File: s3backgrounddelete/ut/cortx_s3_iem_tests.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        event_id = 100
+        message_blob = "*******This is a test IEM message********"
+
+        result_data = s3iem.send(module, event_id, severity, message_blob)
+        assert result_data == True
+    else:
+        assert True == True
+
+def test_receive():
+    if os.path.isfile('/etc/cortx/cluster.conf') and os.stat("/etc/cortx/cluster.conf").st_size > 10:
```
Comment (vimalraghwani20) [inline s3backgrounddelete/ut/cortx_s3_iem_tests.py:38]: Add TODO. we can validate topic IEM

Comment (pranavPawar07) [reply s3backgrounddelete/ut/cortx_s3_iem_tests.py:38]: noted.

**Review by vimalraghwani20 — APPROVED** [review]

**Comment by pranavPawar07** [comment]
Pre merge : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1626/console

**Pull request merged**